lib/recursive-methods.rb

```diff
@@ -1,49 +1,72 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1 
+  rev_s = reverse(s[1..-1])
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.